File: elements-of-programming-interviews/cpp/even_odd_array.cc
#include <set>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::vector;

bool cmp(const int& a, const int& b) {
  if ((a & 1) == 0 && (b & 1) != 0) return true;   // a is even, b is odd
  if ((a & 1) != 0 && (b & 1) == 0) return false;  // a is odd, b is even
  return false;  // either both are even or both are odd; maintain original
                 // order
}

void EvenOddSort(vector<int>* A_ptr) {
  // Start with l = 0, r = len-1
  // sort, but comparator is just "equal less than odd";
  // two odd numbers are sorted, and two even numbers are sorted
  std::sort(A_ptr->begin(), A_ptr->end(), cmp);
  return;
}

void printArr(vector<int>& A) {
  printf("\n[");
  for (auto val : A) {
    printf("%d ", val);
  }
  printf("]\n");
}

void EvenOdd(vector<int>* A_ptr) {
  vector<int>& A = *A_ptr;
  size_t l = 0, r = A.size() - 1;
  while (l < r) {
    if ((A[l] & 1) && !(A[r] & 1)) {
      int tmp = A[l];
      A[l] = A[r];
      A[r] = tmp;
    }
    if (!(A[l] & 1)) {
      l++;
    }
    if (A[r] & 1) {
      r--;
    }
  }
}

void EvenOddWrapper(TimedExecutor& executor, vector<int> A) {
  std::multiset<int> before(begin(A), end(A));

  executor.Run([&] { EvenOdd(&A); });

  bool in_odd = false;
  for (int a : A) {
    if (a % 2 == 0) {
      if (in_odd) {
        throw TestFailure("Even elements appear in odd part");
      }
    } else {
      in_odd = true;
    }
  }

  std::multiset<int> after(begin(A), end(A));
  if (before != after) {
    throw TestFailure("Elements mismatch");
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "A"};
  return GenericTestMain(args, "even_odd_array.cc", "even_odd_array.tsv",
                         &EvenOddWrapper, DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/2025/epi_judge_python/even_odd_array.py
import collections
import functools
from typing import List

from test_framework import generic_test
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook

"""
[2,4,6,3,5,7]
[1,2,3,4,5,6]
"""

def even_odd_first(A: List[int]) -> None:
    i, j = 0, len(A)-1
    while i < j:
        if A[i] % 2 == 0:
            i += 1
            continue 
        if A[j] % 2:
            j -= 1 
            continue
        if A[i] % 2 and A[j] % 2 == 0:
            A[i], A[j] = A[j], A[i]
            i, j = i+1, j-1
    return

def even_odd(A):
    l, r = 0, len(A)-1
    while l < r:
        if not (A[l] % 2 and (A[r] % 2 == 0)): # If the condition we need for swapping is false
            l += 0 if A[l] % 2 else 1          # increment l if A[l] is even
            r -= 1 if A[r] % 2 else 0          # decrement r if A[r] is odd 
        else:
            A[l], A[r] = A[r], A[l]            # Otherwise, swap and continue 

@enable_executor_hook
def even_odd_wrapper(executor, A):
    before = collections.Counter(A)

    executor.run(functools.partial(even_odd, A))

    in_odd = False
    for a in A:
        if a % 2 == 0:
            if in_odd:
                raise TestFailure('Even elements appear in odd part')
        else:
            in_odd = True
    after = collections.Counter(A)
    if before != after:
        raise TestFailure('Elements mismatch')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('even_odd_array.py',
                                       'even_odd_array.tsv', even_odd_wrapper))
